### PR TITLE
Disallow var syntax in string interpolation

### LINKED ISSUE #32951: Weird error with `$var` and interpolation
```julia
julia> function build_variable(var::Expr)
         throw(error("Expected $var to be a variable name"))
       end
ERROR: syntax: missing comma or ) in argument list
Stacktrace:
 [1] top-level scope at none:0

julia> versioninfo()
Julia Version 1.3.0-rc1.0
Commit 768b25f6a8 (2019-08-18 00:04 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, ivybridge)
```

### PR DESCRIPTION
**Pull request by c42f**
The `var"##"` syntax should be disabled in string interpolation.

Also work around a subtle problem with command interpolation's use of Meta.parse in non-greedy mode, combined with the new `var` syntax which would cause `Base.shell_parse("\"(\$var)\"")` to miss the trailing `)` character.

This is a quick fix for a bit of fallout from #32408 which affects PyCall's build script, among other things.

Fixes #32951

### CODE REVIEW

File: src/julia-parser.scm [hunk 1]
```diff
@@ -2285,10 +2285,14 @@
                       (if (closing-token? t)
                           (error (string "unexpected \"" (take-token s) "\"")))))
            (take-token s)
-           (if (and (eq? t 'var) (eqv? (peek-token s) #\") (not (ts:space? s)))
+           (if (and (eq? t 'var)
+                    (if (not (or (ts:pbtok s) (ts:last-tok s)))
+                      ;; Hack to preserve (io.pos (ts:port s)) for non-greedy Meta.parse
```
Comment (JeffBezanson) [inline src/julia-parser.scm]: Indentation is off here

File: src/julia-parser.scm [hunk 2]
```diff
@@ -2285,10 +2285,14 @@
...
+                      (eqv? (peek-char (ts:port s)) #\"))
+                    (and (eqv? (peek-token s) #\") (not (ts:space? s))))
```
Comment (JeffBezanson) [inline src/julia-parser.scm]: I think it would be clearer to nest this inside the previous `if` statement.

Comment (c42f) [reply src/julia-parser.scm]: This ordering was intentional: we must run `peek-token` before the `take-token` further down.  However, we musn't run `peek-token` unless necessary.

It's delicate I'm afraid.

Perhaps I'll just add an extra `peek-token` down below and rearrange as suggested. That would be clearer if (very slightly) redundant.

File: src/julia-parser.scm (hunk 1 above)
Comment (c42f) [reply src/julia-parser.scm]: Oops thanks. I've adjusted my autoindent settings.

**Comment by c42f** [comment]
Right I've sorted those problems out!

I also added a workaround and test to preserve the behavior of `Base.shell_parse` (and hence `@cmd` interpolation) with `$var` interpolation. This is ugly but the alternative is a non backward compatible change to the way that `@cmd` handles `$` interpolation in general.

I think we *should* change the way `$` interpolation works in `cmd`, but I don't want to couple that breaking change with getting this bug fixed.

**Comment by c42f** [comment]
CI errors look unrelated (win64 — temp files; freebsd — an internal build slave error).

**Comment by c42f** [comment]
@JeffBezanson can you live with the ugly `shell_parse` compatibility workaround I've got here? If so I'd like to merge this and come back to the inconsistency between cmd interpolation vs string interpolation another time.

**Comment by MikeInnes** [comment]
This bug breaks both Juno and the CUDA stack, so it'd be great to get the fix in quickly.

**Comment by chriscoey** [comment]
bump

**Pull request merged**